Lib9c/Renderer/ValidatingActionRenderer.cs

```diff
@@ -92,51 +89,105 @@ public override void RenderBlockEnd(Block<T> oldTip, Block<T> newTip)
         {
             base.RenderBlockEnd(oldTip, newTip);
             Validate();
+        }
 
-            List<IAction> expectedActions = new List<IAction>();
+        public override void RenderReorgEnd(
+            Block<T> oldTip,
+            Block<T> newTip,
+            Block<T> branchpoint
+        )
+        {
+            base.RenderReorgEnd(oldTip, newTip, branchpoint);
+            Validate();
+
+            ValidateReorgEnd(Records, oldTip, newTip, branchpoint);
+        }
+
+        private void ValidateReorgEnd(
+            IReadOnlyList<RenderRecord<T>> records,
+            Block<T> oldTip,
+            Block<T> newTip,
+            Block<T> branchpoint)
+        {
+            var store = Store ?? throw new Exception("Store is null.");
+            var chain = BlockChain ?? throw new Exception("Chain is null.");
             
-            foreach (var tx in newTip.Transactions)
+            List<IAction> expectedUnrenderedActions = new List<IAction>();
+            List<IAction> expectedRenderedActions = new List<IAction>();
+
+            var block = oldTip;
+            bool repeat;
+            do
             {
-                expectedActions.AddRange(tx.Actions.Cast<IAction>());
+                repeat = !block.PreviousHash.Equals(branchpoint.Hash);
+                expectedUnrenderedActions.AddRange(
+                    block.Transactions.SelectMany(t => t.Actions).Cast<IAction>());
+                if (!(chain.Policy.BlockAction is null))
+                {
+                    expectedUnrenderedActions.Add(chain.Policy.BlockAction);
+                }
+
+                block = block.PreviousHash is null ? chain.Genesis : store.GetBlock<T>(block.PreviousHash.Value);
```

Comment:
I guess `block` is already `chain.Genesis` if `block.PreviousHash is null`. 😕

Reply:
You are right. Amended in [`9e44d67`](https://github.com/planetarium/lib9c/pull/124/commits/9e44d67e40b808fda503a4f58e86a3bf96dae280).